File: APM/Java/Interpreter/src/ro/sandorrobertk94/view/ConsoleUI.java
package ro.sandorrobertk94.view;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import ro.sandorrobertk94.controller.Controller;
import ro.sandorrobertk94.domain.ProgramState;
import ro.sandorrobertk94.domain.adts.MyDictionary;
import ro.sandorrobertk94.domain.adts.MyList;
import ro.sandorrobertk94.domain.adts.MyStack;
import ro.sandorrobertk94.domain.expressions.ArithmeticExpression;
import ro.sandorrobertk94.domain.expressions.ConstExpression;
import ro.sandorrobertk94.domain.expressions.Expression;
import ro.sandorrobertk94.domain.expressions.VariableExpression;
import ro.sandorrobertk94.domain.statements.AssignmentStatement;
import ro.sandorrobertk94.domain.statements.CompoundStatement;
import ro.sandorrobertk94.domain.statements.IStatement;
import ro.sandorrobertk94.domain.statements.IfStatement;
import ro.sandorrobertk94.domain.statements.PrintStatement;
import ro.sandorrobertk94.domain.statements.WhileStatement;

/**
 * Created by Robert on 10/26/2015.
 */
public class ConsoleUI {
    private Controller controller;

    public ConsoleUI(Controller controller) {
        this.controller = controller;
    }

    public void run() {
        while (true) {
            printMainMenu();

            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            int option = 0;
            try {
                option = Integer.parseInt(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }

            switch (option) {
                case 1: {
                    ProgramState state = inputProgram();
                    if (state == null) {
                        return;
                    }
                    controller.getRepo().add(state);
                    System.out.println();
                    System.out.println(state.getOriginalProgram().toString());
                    System.out.println();
                    break;
                }
                case 2: {
                    oneStep();
                    break;
                }
                case 3: {
                    allStep();
                    break;
                }
                case 4: {
                    return;
                }
                default: {
                    System.out.println("!! Invalid Option !!");
                }
            }
        }
    }

    private ProgramState inputProgram() {
        IStatement initialStatement;
        MyStack initialStack = new MyStack();

        initialStatement = inputStatement();
        if (initialStatement == null) {
            return null;
        }

        initialStack.push(initialStatement);

        return new ProgramState(
                initialStack,
                new MyDictionary(),
                new MyList(),
                initialStatement
        );
    }

    private IStatement inputStatement() {
        printStatementMenu();
        IStatement statement;

        // read option
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int option = 0;
        try {
            option = Integer.parseInt(br.readLine());
        } catch (IOException e) {
            e.printStackTrace();
        }

        switch (option) {
            case 1: {
                statement = new CompoundStatement();

                System.out.println("Input first statement : ");
                IStatement st1 = inputStatement();
                System.out.println("Input second statement : ");
                IStatement st2 = inputStatement();

                if (st1 == null || st2 == null) {
                    return null;
                }

                ((CompoundStatement) statement).setFirstStatement(st1);
                ((CompoundStatement) statement).setSecondStatement(st2);
                break;
            }
            case 2: {
                statement = new AssignmentStatement();

                System.out.print("Variable id : ");
                String variableid = null;
                try {
                    variableid = br.readLine();
                } catch (IOException e) {
                    e.printStackTrace();
                }

                System.out.println("Input expression : ");
                Expression exp = inputExpression();
                if (exp == null) {
                    return null;
                }

                ((AssignmentStatement) statement).setVariableID(variableid);
                ((AssignmentStatement) statement).setExp(exp);
                break;
            }
            case 3: {
                statement = new IfStatement();

                System.out.println("Input expression : ");
                Expression exp = inputExpression();
                System.out.println("Input first statement : ");
                IStatement st1 = inputStatement();
                System.out.println("Input second statement : ");
                IStatement st2 = inputStatement();

                if (exp == null || st1 == null || st2 == null) {
                    return null;
                }

                ((IfStatement) statement).setExp(exp);
                ((IfStatement) statement).setThenStatement(st1);
                ((IfStatement) statement).setElseStatement(st2);
                break;
            }
            case 4: {
                statement = new PrintStatement();
                System.out.println("Input expression : ");
                Expression exp = inputExpression();
                if (exp == null) {
                    return null;
                }
                ((PrintStatement) statement).setExp(exp);
                break;
            }
            case 5: {
                statement = new WhileStatement();
                System.out.println("Input expression : ");
                Expression exp = inputExpression();
                System.out.println("Input statement : ");
                IStatement st = inputStatement();
                if (exp == null || st == null) {
                    return null;
                }
                ((WhileStatement) statement).setExp(exp);
                ((WhileStatement) statement).setStatement(st);
            }
            default: {
                System.out.println("!! Invalid Option !!");
                return null;
            }
        }

        return statement;
    }

    private Expression inputExpression() {
        printExpressionMenu();

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int option = 0;
        try {
            option = Integer.parseInt(br.readLine());
        } catch (IOException e) {
            e.printStackTrace();
        }

        Expression exp;
        switch (option) {
            case 1: {
                System.out.print("Input an integer : ");
                int number = 0;
                try {
                    number = Integer.parseInt(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }

                exp = new ConstExpression(number);
                break;
            }
            case 2: {
                exp = new ArithmeticExpression();

                System.out.println("Input expression : ");
                Expression exp1 = inputExpression();

                System.out.print("Input operator ( + , - , * , / ) : ");
                String sop = null;
                try {
                    sop = br.readLine();
                } catch (IOException e) {
                    e.printStackTrace();
                }

                if (sop != null) {
                    switch (sop) {
                        case "+": {
                            ((ArithmeticExpression) exp).setOp(ArithmeticExpression.Operator.ADD);
                            break;
                        }
                        case "-": {
                            ((ArithmeticExpression) exp).setOp(ArithmeticExpression.Operator.SUB);
                            break;
                        }
                        case "*": {
                            ((ArithmeticExpression) exp).setOp(ArithmeticExpression.Operator.MULT);
                            break;
                        }
                        case "/": {
                            ((ArithmeticExpression) exp).setOp(ArithmeticExpression.Operator.DIV);
                            break;
                        }
                        default: {
                            System.out.println("!! Bad Operator !!");
                            return null;
                        }
                    }
                }

                System.out.println("Input expression : ");
                Expression exp2 = inputExpression();

                if (exp1 == null || exp2 == null) {
                    return null;
                }

                ((ArithmeticExpression) exp).setExp1(exp1);
                ((ArithmeticExpression) exp).setExp2(exp2);
                break;
            }
            case 3: {
                exp = new VariableExpression();

                System.out.print("Variable id :");
                String variableid = null;
                try {
                    variableid = br.readLine();
                } catch (IOException e) {
                    e.printStackTrace();
                }

                ((VariableExpression) exp).setVariableID(variableid);
                break;
            }
            default: {
                System.out.println("!! Invalid Option !!");
                return null;
            }
        }
        return exp;
    }

    private void oneStep() {
        String output = controller.oneStep(controller.getRepo().getCurrentProgram(0));
        System.out.println(output);
    }

    private void allStep() {
        String output = controller.allStep();
        System.out.println(output);
    }

    private void printMainMenu() {
        System.out.println("--- Main Menu ---");
        System.out.println("1. Input program");
        System.out.println("2. One Step");
        System.out.println("3. All Step");
        System.out.println("4. Exit");
    }

    private void printStatementMenu() {
        System.out.println("1. Compound Statement");
        System.out.println("2. Assignment");
        System.out.println("3. If Statement");
        System.out.println("4. Print Statement");
        System.out.println("5. While Statement");
    }

    private void printExpressionMenu() {
        System.out.println("1. Constant Expression");
        System.out.println("2. Arithmetic Expression");
        System.out.println("3. Variable Expression");
    }
}

File: APM/CSharp/Interpreter/Interpreter/view/ConsoleUI.cs
using System;
using Interpreter.controller;
using Interpreter.domain;

namespace Interpreter.view
{
    public class ConsoleUI
    {
        private readonly Controller _controller;

        public ConsoleUI(Controller controller)
        {
            _controller = controller;
        }

        public void Run()
        {
            while (true)
            {
                PrintMainMenu();

                var option = Convert.ToInt32(Console.ReadLine());

                switch (option)
                {
                    case 1:
                    {
                        var state = InputProgram();
                        if (state == null)
                        {
                            return;
                        }

                        _controller.Repo.Add(state);
                        Console.WriteLine();
                        Console.WriteLine(state.OriginalProgram.ToString());
                        Console.WriteLine();
                        break;
                    }
                    case 2:
                    {
                        OneStep();
                        break;
                    }
                    case 3:
                    {
                        AllStep();
                        break;
                    }
                    case 4:
                    {
                        return;
                    }
                    default:
                    {
                        Console.WriteLine("!!! Invalid option !!!");
                        break;
                    }
                }
            }
        }

        private ProgramState InputProgram()
        {
            var initialStack = new MyStack();

            var initialStatement = InputStatement();
            if (initialStatement == null)
            {
                return null;
            }

            initialStack.Push(initialStatement);
            return new ProgramState(initialStack, new MyDictionary(), new MyList(), initialStatement);
        }

        private IStatement InputStatement()
        {
            PrintStatementMenu();
            IStatement statement;

            var option = Convert.ToInt32(Console.ReadLine());

            switch (option)
            {
                case 1:
                {
                    statement = new CompoundStatement();

                    Console.WriteLine("Input first statement : ");
                    var st1 = InputStatement();
                    Console.WriteLine("Input second statement : ");
                    var st2 = InputStatement();

                    if (st1 == null || st2 == null)
                    {
                        return null;
                    }

                    ((CompoundStatement) statement).FirstStatement = st1;
                    ((CompoundStatement) statement).SecondStatement = st2;
                    break;
                }
                case 2:
                {
                    statement = new AssignmentStatement();

                    Console.WriteLine("Variable id :");
                    var varid = Console.ReadLine();

                    Console.WriteLine("Input exception");
                    var exp = InputExpression();
                    if (exp == null)
                    {
                        return null;
                    }

                    ((AssignmentStatement) statement).VariableId = varid;
                    ((AssignmentStatement) statement).Exp = exp;
                    break;
                }
                case 3:
                {
                    statement = new IfStatement();
                    Console.WriteLine("Input expression : ");
                    var exp = InputExpression();
                    Console.WriteLine("Input then statement : ");
                    var st1 = InputStatement();
                    Console.WriteLine("Input else statement : ");
                    var st2 = InputStatement();

                    if (exp == null || st1 == null || st2 == null)
                    {
                        return null;
                    }

                    ((IfStatement) statement).Exp = exp;
                    ((IfStatement) statement).ThenStatement = st1;
                    ((IfStatement) statement).ElseStatement = st2;
                    break;
                }
                case 4:
                {
                    statement = new PrintStatement();

                    Console.WriteLine("Input expression : ");
                    var exp = InputExpression();
                    if (exp == null)
                    {
                        return null;
                    }
                    ((PrintStatement) statement).Exp = exp;
                    break;
                }
                default:
                {
                    Console.WriteLine("Invalid option!");
                    return null;
                }
            }
            return statement;
        }

        private Expression InputExpression()
        {
            PrintExpressionMenu();

            var option = Convert.ToInt32(Console.ReadLine());
            Expression exp;
            switch (option)
            {
                case 1:
                {
                    Console.Write("Input an integer : ");
                    var number = Convert.ToInt32(Console.ReadLine());
                    exp = new ConstantExpression(number);
                    break;
                }
                case 2:
                {
                    exp = new ArithmeticExpression();

                    Console.WriteLine("Input expression : ");
                    var exp1 = InputExpression();

                    Console.Write("Input operator ( + , - , * , / ) : ");
                    var sop = Console.ReadLine();
                    switch (sop)
                    {
                        case "+":
                        {
                            ((ArithmeticExpression) exp).Op = ArithmeticExpression.Operator.Add;
                            break;
                        }
                        case "-":
                        {
                            ((ArithmeticExpression) exp).Op = ArithmeticExpression.Operator.Sub;
                            break;
                        }
                        case "*":
                        {
                            ((ArithmeticExpression) exp).Op = ArithmeticExpression.Operator.Mul;
                            break;
                        }
                        case "/":
                        {
                            ((ArithmeticExpression) exp).Op = ArithmeticExpression.Operator.Div;
                            break;
                        }
                        default:
                        {
                            Console.WriteLine("!!! Bad Operator !!!");
                            return null;
                        }
                    }

                    Console.WriteLine("Input expression : ");
                    var exp2 = InputExpression();

                    if (exp1 == null || exp2 == null)
                    {
                        return null;
                    }
                    ((ArithmeticExpression) exp).Exp1 = exp;
                    ((ArithmeticExpression) exp).Exp2 = exp2;
                    break;
                }
                case 3:
                {
                    Console.Write("Variable id : ");
                    var varid = Console.ReadLine();
                    exp = new VariableExpression(varid);
                    break;
                }
                default:
                {
                    Console.WriteLine("!!! Invalid option !!!");
                    return null;
                }
            }
            return exp;
        }

        private void OneStep()
        {
            var output = _controller.OneStep(_controller.Repo.GetCurentProgramState(0));
            Console.WriteLine(output);
        }

        private void AllStep()
        {
            var output = _controller.AllStep();
            Console.WriteLine(output);
        }

        private void PrintMainMenu()
        {
            Console.WriteLine("--- Main Menu ---");
            Console.WriteLine("1. Input program");
            Console.WriteLine("2. One Step");
            Console.WriteLine("3. All Step");
            Console.WriteLine("4. Exit");
        }

        private void PrintStatementMenu()
        {
            Console.WriteLine("1. Compound Statement");
            Console.WriteLine("2. Assignment");
            Console.WriteLine("3. If Statement");
            Console.WriteLine("4. Print Statement");
        }

        private void PrintExpressionMenu()
        {
            Console.WriteLine("1. Constant Expression");
            Console.WriteLine("2. Arithmetic Expression");
            Console.WriteLine("3. Variable Expression");
        }
    }
}